lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -0,0 +1,809 @@
+    if (Arg2AllocStackMap.count(ArgIndex) > 0) {
+      auto *ASI = dyn_cast<AllocStackInst>(Arg2AllocStackMap[ArgIndex]);
+      if (ExistentialArgDescriptor[ArgIndex].DestroyAddrUse) {
+        for (Operand *ASIUse : ASI->getUses()) {
+          auto *ASIUser = ASIUse->getUser();
+          if (auto *DAI = dyn_cast<DestroyAddrInst>(ASIUser)) {
+            SILBuilder Builder(ASIUser);
+            Builder.setInsertionPoint(&*std::next(ASIUser->getIterator()));
+            Builder.createDeallocStack(DAI->getLoc(), ASI);
+          }
+        }
+      } else { // Need to insert DeallocStack before return.
+        for (auto *I : ReturnInsts) {
+          SILBuilder Builder(I->getParent());
+          Builder.setInsertionPoint(I);
+          Builder.createDeallocStack(ASI->getLoc(), ASI);
+        }
+      }
+    }
+  }
+}
+
+/// Create the Thunk Body with always_inline attribute.
+void ExistentialSpecializerTransform::populateThunkBody() {
+
+  SILModule &M = F->getModule();
+
+  F->setThunk(IsThunk);
+  F->setInlineStrategy(AlwaysInline);
+
+  /// Create a basic block and the function arguments.
+  SILBasicBlock *ThunkBody = F->createBasicBlock();
+  for (auto &ArgDesc : ArgumentDescList) {
+    ThunkBody->createFunctionArgument(ArgDesc.Arg->getType(), ArgDesc.Decl);
+  }
+
+  /// Builder to add new instructions in the Thunk.
+  SILBuilder Builder(ThunkBody);
+  SILOpenedArchetypesTracker OpenedArchetypesTracker(F);
+  Builder.setOpenedArchetypesTracker(&OpenedArchetypesTracker);
+  Builder.setCurrentDebugScope(ThunkBody->getParent()->getDebugScope());
+
+  /// Location to insert new instructions.
+  SILLocation Loc = ThunkBody->getParent()->getLocation();
+
+  /// Create the function_ref instruction to the NewF.
+  FunctionRefInst *FRI = Builder.createFunctionRef(Loc, NewF);
+
+  auto GenCalleeType = NewF->getLoweredFunctionType();
+  auto CalleeGenericSig = GenCalleeType->getGenericSignature();
+
+  /// Determine arguments to Apply.
+  /// Generate opened existentials for generics.
+  llvm::SmallVector<SILValue, 8> ApplyArgs;
+
+  /// Ideally this should be a SubstitutionMap. TODO.
+  llvm::DenseMap<SubstitutableType *, Type> Generic2OpenedTypeMap;
+  for (auto &ArgDesc : ArgumentDescList) {
+    if (Arg2GenericTypeMap[ArgDesc.Index]) {
+      ArchetypeType *Opened;
+      SILValue OrigOperand = ThunkBody->getArgument(ArgDesc.Index);
+      auto SwiftType = ArgDesc.Arg->getType().getASTType();
+      auto OpenedType =
+          SwiftType->openAnyExistentialType(Opened)->getCanonicalType();
+      auto OpenedSILType = NewF->getModule().Types.getLoweredType(OpenedType);
+      SILValue archetypeValue;
+      auto ExistentialRepr =
+          ArgDesc.Arg->getType().getPreferredExistentialRepresentation(M);
+      switch (ExistentialRepr) {
+      case ExistentialRepresentation::Opaque: {
+        archetypeValue = Builder.createOpenExistentialAddr(
+            Loc, OrigOperand, OpenedSILType, OpenedExistentialAccess::Mutable);
```

Comment:
Does it have to be mutable? You're not specializing 'inout ExistentialType' to 'inout T' are you? For CoW existentials, mutable existentials incur a copy of the box so you don't want to do it unless you need to